### PR TITLE
commented fft_mt_r2iq ..

### PR DESCRIPTION
**Pull request by hayguen**
.. and it's core, the fast convolution with decimation

* peeled out the core algorithm and blocks with braces and comments
* some performance hints marked with '@todo'
* removed getDecimate() and getFftN() - each used at one single place

### CODE REVIEW

**Comment by hayguen** [comment]
> Remove a pair of ()

force push contains this minor change

File: Core/fft_mt_r2iq.cpp [hunk 1]
```diff
@@ -310,45 +336,76 @@ void * fft_mt_r2iq::r2iqThreadf(r2iqThreadArg *th) {
...
+				// 'full' transformation size: 2 * halfFft
+				fftwf_execute_dft_r2c(th->plan_t2f_r2c, th->ADCinTime + (3 * halfFft / 2) * k, th->ADCinFreq);
+				// result now in th->ADCinFreq[]
 
-			for (int m = start; m < mfft / 2; m++) // circular shift tune fs/2 second half array
-			{
-				dest[m][0] = (source2[m][0] * filter2[m][0] +
-								source2[m][1] * filter2[m][1]);
+				// circular shift (mixing in full bins) and low/bandpass filtering (complex multiplication)
+				{
```
Comment (howard0su) [inline Core/fft_mt_r2iq.cpp:359]: Can we optimize here if decimate = 0, then we don't need a digit filter?

File: Core/fft_mt_r2iq.h [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 #include "r2iq.h"
 #include "fftw3.h"
+
+// use up to this many threads - really that many? (HA)
 #define N_R2IQ_THREAD 64
```
Comment (howard0su) [inline Core/fft_mt_r2iq.h:7]: my laptop show 4 threads are more than enough.

File: Core/fft_mt_r2iq.cpp (hunk 1 above)
Comment (hayguen) [reply Core/fft_mt_r2iq.cpp:359]: From bandwidth perspective, you could be right .. but, with really unsure about this.
We could try to derive this from formulas .. of overlap-scrap .. or simply try-out!

Comment (hayguen) [reply Core/fft_mt_r2iq.cpp:359]: i fear we get a very very low stopband attenuation of 13 dB for the rectangle window ..

Comment (howard0su) [reply Core/fft_mt_r2iq.cpp:359]: so 120db here is just fooling ourselves?

File: Core/fft_mt_r2iq.h (hunk 2 above)
Comment (hayguen) [reply Core/fft_mt_r2iq.h:7]: i have same opionion .. and fear some mess if operating system might change thread/core association - just some paranoia

File: Core/fft_mt_r2iq.cpp (hunk 1 above)
Comment (hayguen) [reply Core/fft_mt_r2iq.cpp:359]: we are now mixing things
* 120 dB is surely some fooling .. cause the Kaiser window function won't be able to create such a good filter with such few coefficients in relation to fft size **for high decimation**
* kaiser window is applied/integrated into the low/bandpass filter coefficients .. if we optimize for decimation=0 by removing the complex multiplication with filter2[] .. we remove the windowing .. and end up with sort of rectangle

Comment (hayguen) [reply Core/fft_mt_r2iq.cpp:359]: > Can we optimize here if decimate = 0, then we don't need a digit filter?

putting the for loops with complex multiplication into an if .. replacing the complex multiplication by simple copy for decimate==0 .. still looks quite good on VHF - except the attenuation is really missing!, which ist sort of integrated in the filter coefficients. thus, it's really necessary to test on HF without the R820T's filters covering problems

File: Core/fft_mt_r2iq.h (hunk 2 above)
Comment (ik1xpv) [reply Core/fft_mt_r2iq.h:7]: Yes. My old I7-3770 shows 4 cores and 8 threads limit :-)

Comment (hayguen) [reply Core/fft_mt_r2iq.h:7]: @ik1xpv : i'm not sure if i understood your comment correct. would you suggest `#define N_R2IQ_THREAD 4` or not?

Comment (howard0su) [reply Core/fft_mt_r2iq.h:7]: this is not actual thread. this is max threads. the thread number is min(64, # of cores). I suggest 4 here.

Comment (hayguen) [reply Core/fft_mt_r2iq.h:7]: @howard0su yes, i knew. rename N_R2IQ_THREAD to N_MAX_R2IQ_THREADS ?

Comment (howard0su) [reply Core/fft_mt_r2iq.h:7]: fine with me.

Comment (hayguen) [reply Core/fft_mt_r2iq.h]: renamed and reduced to 4

Comment (ik1xpv) [reply Core/fft_mt_r2iq.h]: > @ik1xpv : i'm not sure if i understood your comment correct. would you suggest `#define N_R2IQ_THREAD 4` or not?

> @hayguen : I made on my PC  a run with  N_MAX_R2IQ_THREADS 4, 8 or 64.
I'm not able to see difference in CPU load. It seems to me 4 is enough.

**Review by howard0su — APPROVED** [review]

**Review by ik1xpv — APPROVED** [review]

**Pull request merged**